### PR TITLE
feat: match array properties element-wise for non-set operators

### PR DESCRIPTION
**Pull request by kyeh-amp**
### Summary

Adds element-wise (any-match) evaluation for multi-valued user properties when paired with non-set operators (`is`, `contains`, `greater`, etc.), aligning the local evaluation engine with Amplitude analytics/charts behavior.

Previously, non-set operators always stringified the whole property value via `coerce_string` — so an array property like `["a","b"]` would never satisfy `is "a"`. Now:

- `match_condition` calls `coerce_string_array` for every non-null value. If the value is multi-valued (native array or JSON array string) and the operator is non-set, it delegates to a new `match_strings_non_set` helper that returns true if **any** element satisfies the operator.
- Negation operators (`is not`, `does not contain`) also use any-match semantics — `is not "A"` on `["A","B"]` is `true` because `"B"` is not `"A"`.
- Scalars still fall through to the existing `coerce_string` → `match_string` path unchanged.

Also fixes two issues in `coerce_string_array`:

- Added a `start_with?('[')` pre-check so scalar strings skip `JSON.parse` and avoid exception-driven control flow on every evaluation.
- Corrected a pre-existing bug where scalars (non-array JSON or malformed JSON) were returned as `[value]` instead of `nil`, which caused set operators to spuriously match single-string properties.

### Tests

- New `spec/experiment/evaluation/evaluation_spec.rb` with an inline test harness (`flag_with_condition`, `context_with_prop`, `evaluate`, `assert_match`, `assert_no_match`) and 14 unit cases covering scalars, native arrays, JSON array strings, malformed JSON, and leading-whitespace edge cases.
- 7 new integration cases added inline in the existing `operator tests` block; the deployment key is consolidated to `server-VVhLULXCxxY0xqmszXouXxiEzoeJWmSh` (a superset of the prior key, containing all existing flags plus the new multi-value flags).

Full suite: 271 examples, 0 failures. Rubocop: clean.

### Checklist

- [x] Does your PR title have the correct [title format](https://github.com/amplitude/experiment-ruby-server/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
- Does your PR have a breaking change?: No

---

> [!NOTE]
> **Medium Risk**
> Changes core flag targeting semantics when a property is an array/JSON array string, which can alter which users qualify for segments. Also tightens `coerce_string_array` parsing behavior, potentially changing outcomes for previously mis-coerced scalar values.
> 
> **Overview**
> The evaluation engine now treats multi-valued properties (native arrays or JSON array strings) as **element-wise** inputs for *non-set* operators like `is`, `contains`, and comparisons, matching if **any** element satisfies the operator (via new `match_strings_non_set`).
> 
> `coerce_string_array` is tightened to only parse JSON when the string looks like an array (starts with `'['`) and to return `nil` for non-array/malformed JSON instead of coercing scalars into single-element arrays, reducing accidental set-operator matches. Tests are expanded with new unit coverage and additional integration cases, plus the integration spec caches the flags URI and updates the deployment key.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c73b5c6c0db9b0b50f1ced2cc8e1d2bc4de82393. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by vaibhav-jain-exp — APPROVED** [review]

File: lib/experiment/evaluation/evaluation.rb [hunk 1]
```diff
@@ -66,24 +66,25 @@ def evaluate_conditions(target, conditions)
...
-          else
-            false
-          end
+          prop_value_string ? match_string(prop_value_string, condition.op, condition.values) : false
         end
       end
 
+      def match_strings_non_set(prop_values, op, filter_values)
+        prop_values.any? { |v| match_string(v, op, filter_values) }
+      end
```
[the remark this answers is not in the document]
Comment (kyeh-amp) [reply lib/experiment/evaluation/evaluation.rb:86]: see https://github.com/amplitude/experiment-python-server/pull/74#issuecomment-4300136408
